src/nnpack/libnnpack.jl

```diff
@@ -0,0 +1,135 @@
+#NOTE: We do the error handling of nnp_initialize while loading NNPACK
+function nnp_initialize()
+    ccall((:nnp_initialize, libnnpack), nnp_status, (),)
+end
+
+function nnp_deinitialize()
+    @check ccall((:nnp_deinitialize, libnnpack), nnp_status, (),)
+end
+
+function pthreadpool_create(n = 0)
+    ccall((:pthreadpool_create, libnnpack), Ptr{Cvoid}, (Csize_t,), n)
+end
+
+function nnp_relu_output(batch_size, channels, input, output, negative_slope, threadpool)
+    @check ccall((:nnp_relu_output, libnnpack), nnp_status, (Csize_t, Csize_t, Ptr{Cfloat}, Ptr{Cfloat}, Cfloat, pthreadpool_t), batch_size, channels, input, output, negative_slope, threadpool)
+end
+
+function nnp_relu_output(x::Array{Float32,N}, y::Array{Float32,N}; negative_slope::AbstractFloat = 0.0, threadpool = shared_threadpool[]) where {N}
+    # Investigate why the channel and batch dims need to specified like this
+    nnp_relu_output(prod(size(x)[N-1:N]), prod(size(x)[1:N-2]), x, y, negative_slope, threadpool)
+    y
+end
+
+function nnp_relu_input_gradient(batch_size, channels, grad_output, input, grad_input, negative_slope, threadpool)
+    @check ccall((:nnp_relu_input_gradient, libnnpack), nnp_status, (Csize_t, Csize_t, Ptr{Cfloat}, Ptr{Cfloat}, Ptr{Cfloat}, Cfloat, pthreadpool_t), batch_size, channels, grad_output, input, grad_input, negative_slope, threadpool)
+end
+
+function nnp_relu_input_gradient(x::Array{Float32,N}, dy::Array{Float32,N}, dx::Array{Float32,N}; negative_slope::AbstractFloat = 0.0, threadpool = shared_threadpool[]) where {N}
+    # Investigate why the channel and batch dims need to specified like this
+    nnp_relu_input_gradient(Csize_t(prod(size(x)[N-1:N])), prod(size(x)[1:N-2]), dy, x, dx, negative_slope, threadpool)
+    dx
+end
+
+function nnp_softmax_output(batch_size, channels, input, output, threadpool)
+    @check ccall((:nnp_softmax_output, libnnpack), nnp_status, (Csize_t, Csize_t, Ptr{Cfloat}, Ptr{Cfloat}, pthreadpool_t), batch_size, channels, input, output, threadpool)
+end
+
+function nnp_softmax_output(x::VecOrMat{Float32}, y::VecOrMat{Float32}; threadpool = shared_threadpool[])
+    nnp_softmax_output(ndims(x) == 2 ? size(x, 2) : 1, size(x, 1), x, y, threadpool)
+    y
+end
+
+#FIXME: Output of fully connected not consistent with `kernel * input`
```

Comment:
Is this still the case?  If not, please delete these comments.

Reply:
It is not done. I was focussing on getting the convolutions right first before fixing it.